File: Cirque/Cirque/AboutView.swift
//
//  AboutView.swift
//  Cirque
//
//  Created by Nathan Hadley on 5/13/24.
//

import SwiftUI

struct BulletText: View {
    let text: String
    
    var body: some View {
        HStack(alignment: .top) {
            Text("•")
                .font(.body)
                .foregroundColor(.primary)
            Text(text)
                .font(.body)
                .foregroundColor(.primary)
        }
        .padding(.vertical, 2)
    }
}

struct AboutView: View {
    @StateObject private var offlineMapDownloader = OfflineMapDownloader()
    
    let gitHubText = "The code for this project can be found at [GitHub](https://github.com/nathan-hadley/cirque-ios)."
    let circuits = [
        "Forestland Blue Circuit (V0-2)",
        "Swiftwater Red Circuit (V0-3)",
        "Forestland Black Circuit (V3-5)",
        "Straightaways White Circuit (V4-9)"
    ]
    
    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 10) {
                VStack(alignment: .leading, spacing: 10) {
                    Text("About")
                        .font(.largeTitle)
                        .padding(.top, 20)
                    
                    Text("After a trip to Fontainebleau, France, we were inspired to bring their concept of bouldering circuits to Leavenworth. These circuits have been developed so far:")
                    
                    ForEach(circuits, id: \.self) { circuit in
                        BulletText(text: circuit)
                    }
                    
                    Text("If you're a developer, please reach out about contributing. If you're not a developer but want to help, please also reach out. Collecting all the information to add a new circuit takes time!")
                    
                    Text(.init(gitHubText))
                    
                    Text("How to Use")
                        .font(.title)
                        .padding(.top, 20)
                    
                    Text("The app caches map data after viewing for an undetermined amount of time. To ensure circuits show up without network connection, zoom in to the circuit you want to climb before going into the canyon. Stable offline access will be added soon.")
                        .padding(.bottom, 20)
                    
                    Text("Download Maps for Offline Use")
                        .font(.title)
                        .padding(.top, 20)
                    
                    if let successMessage = offlineMapDownloader.successMessage {
                        Text(successMessage)
                            .padding()
                            .background(Color.green)
                            .foregroundColor(Color.white)
                            .cornerRadius(4)
                    }
                    
                    if let errorMessage = offlineMapDownloader.errorMessage {
                        Text(errorMessage)
                            .padding()
                            .background(Color.red)
                            .foregroundColor(Color.white)
                            .cornerRadius(4)
                    }
                    
                    Button(action: offlineMapDownloader.updateMapData) {
                        Text(offlineMapDownloader.mapDownloaded ? "Update Map" : "Download Map")
                            .padding()
                            .background(Color.blue)
                            .foregroundColor(Color.white)
                            .cornerRadius(8)
                    }
                }
                .padding(.horizontal)
                .padding(.bottom)
            }
        }
        .navigationBarTitle("About", displayMode: .inline)
    }
}

struct AboutView_Previews: PreviewProvider {
    static var previews: some View {
        AboutView()
    }
}

File: android/app/src/main/java/com/example/cirque/views/AboutView.kt
package com.example.cirque.views

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.text.ClickableText
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.unit.dp
import androidx.navigation.NavController
import com.example.cirque.ui.theme.CirqueTheme

@Composable
fun AboutView(navController: NavController) {
    val circuits = listOf(
        "Forestland Blue Circuit (V0-2)",
        "Swiftwater Red Circuit (V0-3)",
        "Forestland Black Circuit (V2-5)",
        "Straightaways White Circuit (V4-9)"
    )

    val githubText = buildAnnotatedString {
        val text = "The code for this project can be found on GitHub."
        val startIndex = text.indexOf("GitHub")
        val endIndex = startIndex + 6

        append(text)
        addStyle(style = SpanStyle(color = Color.Blue), startIndex, endIndex)
        addStringAnnotation(
            tag = "URL",
            annotation = "https://github.com/nathan-hadley/cirque-ios",
            start = startIndex,
            end = endIndex
        )
    }
    val uriHandler = LocalUriHandler.current

    val bodyStyle = MaterialTheme.typography.bodyLarge
    val bodyModifier = Modifier.padding(vertical = 12.dp)

    CirqueTheme {
        Surface(
            modifier = Modifier.fillMaxSize(),
            color = MaterialTheme.colorScheme.background
        ) {
            Column(
                modifier = Modifier.fillMaxSize().padding(15.dp),
                horizontalAlignment = Alignment.Start,
                verticalArrangement = Arrangement.Top
            ) {
                Text(
                    "About",
                    style = MaterialTheme.typography.displayMedium,
                    modifier = bodyModifier
                )
                Text(
                    "After a trip to Fontainebleau, France, we were inspired to bring their concept of bouldering circuits to Leavenworth. These circuits have been developed so far:",
                    style = bodyStyle,
                    modifier = bodyModifier
                )
                circuits.forEach { circuit ->
                    BulletText(text = circuit)
                }
                Text(
                    "If you're a developer, please reach out about contributing. If you're not a developer but want to help, please also reach out me at @nathanhadley_ on Instagram or Threads. Collecting all the information to add a new circuit takes time and I could use help!",
                    style = bodyStyle,
                    modifier = bodyModifier
                )
                ClickableText(
                    text = githubText,
                    style = bodyStyle,
                    modifier = bodyModifier,
                    onClick = { githubText
                        .getStringAnnotations("URL", it, it)
                        .firstOrNull()?.let { text ->
                            uriHandler.openUri(text.item)
                        }
                    }
                )
            }
        }
    }
}

@Composable
fun BulletText(text: String) {
    Row(
        verticalAlignment = Alignment.Top,
        modifier = Modifier.padding(vertical = 4.dp)
    ) {
        Text(
            "•",
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onBackground
        )
        Spacer(modifier = Modifier.width(8.dp))
        Text(
            text,
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onBackground
        )
    }
}